File: python/src/scipp/core/counts.py
# SPDX-License-Identifier: BSD-3-Clause
# Copyright (c) 2021 Scipp contributors (https://github.com/scipp)
# @author Matthew Andrew

from typing import Union

from .._scipp import core as _cpp
from ._cpp_wrapper_util import call_func as _call_cpp_func


def counts_to_density(x: Union[_cpp.DataArray, _cpp.Dataset],
                      dim: str) -> Union[_cpp.DataArray, _cpp.Dataset]:
    """Converts counts to count density on a given dimension.

    :param x: Data as counts.
    :param dim: Dimension on which to convert.
    :return: Data as count density.
    """
    return _call_cpp_func(_cpp.counts_to_density, x, dim)


def density_to_counts(x: Union[_cpp.DataArray, _cpp.Dataset],
                      dim: str) -> Union[_cpp.DataArray, _cpp.Dataset]:
    """Converts counts to count density on a given dimension.

    :param x: Data as count density.
    :param dim: Dimension on which to convert.
    :return: Data as counts.
    """
    return _call_cpp_func(_cpp.density_to_counts, x, dim)

File: lib/python/counts.cpp
// SPDX-License-Identifier: BSD-3-Clause
// Copyright (c) 2023 Scipp contributors (https://github.com/scipp)
/// @file
#include "scipp/dataset/counts.h"
#include "scipp/dataset/dataset.h"

#include "nanobind.h"

using namespace scipp;
using namespace scipp::dataset;

namespace nb = nanobind;

void init_counts(nb::module_ &m) {
  m.def(
      "counts_to_density",
      [](const Dataset &d, const std::string &dim) {
        return counts::toDensity(d, Dim{dim});
      },
      nb::arg("x"), nb::arg("dim"));

  m.def(
      "counts_to_density",
      [](const DataArray &d, const std::string &dim) {
        return counts::toDensity(d, Dim{dim});
      },
      nb::arg("x"), nb::arg("dim"));

  m.def(
      "density_to_counts",
      [](const Dataset &d, const std::string &dim) {
        return counts::fromDensity(d, Dim{dim});
      },
      nb::arg("x"), nb::arg("dim"));

  m.def(
      "density_to_counts",
      [](const DataArray &d, const std::string &dim) {
        return counts::fromDensity(d, Dim{dim});
      },
      nb::arg("x"), nb::arg("dim"));
}
